### PR TITLE
Enable cross-compiling of HogMaker

### PR DESCRIPTION
**Pull request by tophyr**
## Pull Request Type
- [ ] GitHub Workflow changes
- [ ] Documentation or Wiki changes
- [x] Build and Dependency changes
- [ ] Runtime changes
  - [ ] Render changes
  - [ ] Audio changes
  - [ ] Input changes
  - [ ] Network changes
  - [ ] Other changes

### Description
HogMaker is a build tool that must run on the host. It must not, then, be compiled as part of the overall target build. For builds that _must_ cross-compile, like Android (and eventually iOS/visionOS), we need to be able to reference an already-built HogMaker project.

### Checklist

- [x] I have tested my changes locally and verified that they work as intended.
- [x] I have documented any new or modified functionality.
- [x] I have reviewed the changes to ensure they do not introduce any unnecessary complexity or duplicate code.
- [x] I understand that by submitting this pull request, I am agreeing to license my contributions under the project's license.

### CODE REVIEW

File: tools/CMakeLists.txt [hunk 1]
```diff
@@ -1,7 +1,12 @@
...
-target_include_directories(HogMaker PRIVATE ${PROJECT_BINARY_DIR}/lib)
+if (ANDROID)
+	find_package(HogMaker REQUIRED)
+else()
+	add_executable(
+		HogMaker
+		HogMaker/HogFormat.cpp
+		HogMaker/HogMaker.cpp
+	)
+	export(TARGETS HogMaker FILE "${CMAKE_BINARY_DIR}/HogMakerConfig.cmake")
```
Comment (Lgt2x) [inline tools/CMakeLists.txt:9]: HogMakerConfig.cmake was not included in this PR. And in this else(), we're not targeting Android, which means that HogMaker will not be looked up using `find_package`. So why exporting the target?

File: tools/CMakeLists.txt [hunk 2]
```diff
@@ -1,7 +1,12 @@
-add_executable(
-	HogMaker
-	HogMaker/HogFormat.cpp
-	HogMaker/HogMaker.cpp
-)
-add_dependencies(HogMaker get_git_hash)
-target_include_directories(HogMaker PRIVATE ${PROJECT_BINARY_DIR}/lib)
+if (ANDROID)
+	find_package(HogMaker REQUIRED)
```
Comment (Lgt2x) [inline tools/CMakeLists.txt:2]: Where is the HogMaker package created in case we're building for Android

**Review by Lgt2x — CHANGES_REQUESTED** [review]
I understand the motivation behind this PR: HogMaker's target arch is not necessarily the same as the game's target arch. With CMake, you can only ever load a single compiler toolchain at a time in a project, so you've decided to create a separate project for HogMaker that uses a separate toolchain, not cross-compiling this time.

Why not make HogMaker a separate project for all archs, and find it in the top-level CMakeLists? This way, the target is available to all subdirectories, without find_package in each one, and without conditionals for Android.

**Comment by tophyr** [comment]
@Lgt2x ah, from your comments it's clear I needed to add some documentation about the process. When cross-compiling, there have to be two CMake invocations: the first builds HogMaker using the host toolchain (and exports the `HogMakerConfig.cmake` file) and then the second, using the target toolchain, gets pointed to the first build.
```
# creates HogMakerConfig.cmake
cmake -B builds/host
# creates the host HogMaker binary
cmake --build builds/host --target HogMaker

# now, do the actual android (or target) build
cmake -B builds/target -DCMAKE_TOOLCHAIN_FILE=/path/to/android/toolchain.cmake -DHogMaker_DIR=$(pwd)/builds/host
cmake --build builds/target
```

So - 
> HogMakerConfig.cmake was not included in this PR.
> Where is the HogMaker package created in case we're building for Android

That file gets created by the first CMake invocation

> in this else(), we're not targeting Android, which means that HogMaker will not be looked up using find_package. So why exporting the target?

correct. we have to export it from this one, so that the android build can subsequently reference it.

**Comment by tophyr** [comment]
Conditionals improved and README updated.

**Comment by tophyr** [comment]
ping @Lgt2x for re-review

**Comment by Lgt2x** [comment]
On a second thought, this solution of running cmake twice with different options to load the native toolchain to build HogMaker, and then running the cross-compilation toolchain for the game adds complexity to our build system. I experimented with a solution using `ExternalProject` cmake logic, which allows running two different toolchains in one build for my WASM build, and considers HogMaker a different project. The change to cross compile HogMaker looks like this: https://github.com/Lgt2x/Descent3/commit/db3a886236b495a38ef29acf76e5886a98db7786

Do you think this could work for Android too?

**Comment by tophyr** [comment]
Reworked PR to use @Lgt2x's idea of `ExternalProject_Add` instead of manually prebuilding and then using `find_package`.

**Comment by tophyr** [comment]
@Lgt2x @winterheart how's this one looking review-wise? I'll get the merge conflicts updated tonight, but are there any other structural/functional/stylistic changes you'd like to see?

**Comment by Lgt2x** [comment]
Well I'm fine with the change as I mostly wrote the patch. Is it fine for you @winterheart ?

**Comment by winterheart** [comment]
Look good for me, but I'm not tested it.

**Comment by winterheart** [comment]
Can we try [this approach](https://cmake.org/cmake/help/book/mastering-cmake/chapter/Cross%20Compiling%20With%20CMake.html#running-executables-built-in-the-project) for cross-compile purposes?
This snippet before `add_custom_command` should do trick:

```cmake
if(CMAKE_CROSSCOMPILING)
   find_package(HogMaker)
else()
   # Build HogMaker executable
   add_executable(HogMaker ...)
   # Export executable as target
   export(TARGETS HogMaker FILE
          "${CMAKE_BINARY_DIR}/HogMakerConfig.cmake")
endif()
# Call add_custom_command with HogMaker involved
...
```

And then first build native tools, after that configure cross-compiled environment:

```sh
# Native tools
mkdir build-native; cd build-native
cmake ..
ninja HogMaker
# Cross-platform build
cd ..
mkdir build-cross; cd build-cross
cmake -DCMAKE_TOOLCHAIN_FILE=MyToolchain.cmake \
      -DHogMaker_DIR=~/src/build-native/ ..
make
```


With this approach you'll need to first build host tools and then reuse them on cross-compile environment.

**Comment by tophyr** [comment]
> With this approach you'll need to first build host tools and then reuse them on cross-compile environment.

This was the _original_ approach with this PR, a few months ago. See above: https://github.com/DescentDevelopers/Descent3/pull/465#issuecomment-2187953682

It worked alright, but @Lgt2x noted:

> this solution of running cmake twice with different options to load the native toolchain to build HogMaker, and then running the cross-compilation toolchain for the game adds complexity to our build system.

The currently-proposed solution adds no complexity to the build process.

**Comment by tophyr** [comment]
@Lgt2x this is blocked on your changes requested btw; seems like you're ok with it so just flagging that

**Review by Lgt2x — APPROVED** [review]

**Comment by Lgt2x** [comment]
sorry, all good :)

**Comment by tophyr** [comment]
Looks like #637 has accomplished this and more in the meantime! <3

**Comment by tophyr** [comment]
Actually, I'm gonna reopen this one - @winterheart / @Lgt2x I still think it'd be overall better to use @Lgt2x's approach of treating HogMaker as an external (rather than sub-) project. This would let us eliminate the two-step build process.

I'll get this branch rebased, but in the meantime - thoughts?

**Comment by Lgt2x** [comment]
Hey @tophyr , I'm glad you're back on the project! I'm quite satisfied by the solution we adopted in #637 , I don't think the 2-step build process is a big limitation, and it has proven to be successful for the ARM64 builds introduced in #642 . Does this solution fall short for your use cases? I'm not too keen on refactoring all of it just for the sake of it

**Comment by tophyr** [comment]
It's good to be back again :)

The benefit to me of having CMake treat HogMaker as an external project is that the overall process is better-encapsulated into how Android's build process sees things. I will play around again with getting the two-step build working in Android (I did it once before, I'm pretty sure) to see how easy that'd be.

**Comment by tophyr** [comment]
Ok, I've got this working honestly pretty smoothly with the HogMaker cross-compilation as-is.

I do still think it's a nicer build to encapsulate it, as I like the idea of `linux-cross-arm64` not needing to be "special" in CI, but my Android stuff is no longer blocked on this.

**Comment by tophyr** [comment]
Update: I *am* blocked on this. The Android build additions needed to configure a second, host-only, native build are janky and complex.

**Comment by tophyr** [comment]
Actually, I've just merged this into https://github.com/DescentDevelopers/Descent3/pull/459 - that PR is now ready for review, and this change makes more sense in the context of that one overall.

**Pull request closed without merging**